lib: add callLocklessFlake

MatthewCroughan opened the pull request:
This is essentially a copy of the function of the same name, from
flake-compat. callLocklessFlake is useful when trying to utilise a
flake.nix without a lock file, often for when you want to create a
subflake from within a parent flake.

Co-authored-by: Tom Bereknyei <tomberek@gmail.com>

Taken from: https://github.com/tomberek/relative/blob/0943a113c19e17b5f51c054de5233f2d3c97ce35/flake.nix#L32-L36
Inspired by: https://github.com/edolstra/flake-compat/blob/64a525ee38886ab9028e6f61790de0832aa3ef03/default.nix#L84-L88
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

MatthewCroughan committed:
lib: add callLocklessFlake

This is essentially a copy of the function of the same name, from
flake-compat. callLocklessFlake is useful when trying to utilise a
flake.nix without a lock file, often for when you want to create a
subflake from within a parent flake.

Co-authored-by: Tom Bereknyei <tomberek@gmail.com>
Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

Artturin commented:
Should this be in flake-utils repo instead of here?

MatthewCroughan commented:
@Artturin It would be a real shame to depend on a library like flake-utils in any flake, than to be able to simply say `nixpkgs.lib.callLocklessFlake`. If this isn't merged, I'll just be using it in the `let` block of my `outputs`, like is the case in all of the examples https://github.com/NixOS/templates/blob/6296caa969420047c3de8b336f481106cf7af6f8/c-hello/flake.nix#L8

roberth commented on lib/trivial.nix:
```diff
@@ -13,6 +13,19 @@ rec {
+     Example:
+       callLocklessFlake ./directoryContainingFlake { inherit nixpkgs; }
+  */
+  callLocklessFlake = path: inputs: let
```
```suggestion
  callLocklessFlake = { path, inputs ? {} }: let
```

Make this extensible. For instance, might want to add a `sourceInfo` in the future; who knows what else.

tomberek replied:
this should be almost exactly the api proposed here: https://github.com/NixOS/nix/issues/6228

roberth commented on lib/trivial.nix:
```diff
@@ -13,6 +13,19 @@ rec {
+       callLocklessFlake ./directoryContainingFlake { inherit nixpkgs; }
+  */
+  callLocklessFlake = path: inputs: let
+    r = {outPath = path;} //
```
```suggestion
    self = {outPath = path;} //
```

roberth replied:
> `{outPath = path;}`

Canonicalize `outPath`? Depending on the type, might need something like `builtins.path { name = "source": ... }`.

Not sure, maybe this should be the caller's responsibility. If they want to avoid putting intermediate paths in the store, perhaps we shouldn't.

tomberek replied:
In most cases I'd see this used via a sub-flake in a monorepo or another flake input which would already be canonicalized.

roberth replied:
I agree. Let's not make the mistake again of doing too much.

tomberek replied:
I don't have a strong opinion either way at the moment.

I'm not sure what the behavior is with builtins.path in term of multiple copies. I'm thinking of cases like a monorepo where a subdirectory already exists in the Store somewhere, just not as a top-level store path.

roberth replied:
After some thought, this scenario seems rather serious:

1. user sees that `outPath` isn't copied, uses it for some sources imports, etc
2. user applies `callLocklessFlake` to directory with confidential files, which is safe to do
3. we change `callLocklessFlake` to match other users' expectations of calling `builtins.path`
4. user updates Nixpkgs
5. user unknowingly puts confidential files in the store or binary cache

A similar scenario exists for the opposite direction, where a store path size (and therefore closure size) can increase unknowingly.

It seems like it's a matter of time before someone comes along to make it do the "right" thing and enact (3), so let's preempt that by adding a `copy` parameter without a default, fixing the problem at (1). We can assign a default once flakes stabilize and when we know how this function is used in practice.

> I'm not sure what the behavior is with builtins.path in term of multiple copies.

Subflakes always call it, so `outPath` is always a top-level path. It may not always do this in the future, but for now it seems that we should replicate that behavior. (Maybe I've been holding it wrong, or maybe this changes)
There's no need for the flake to be a subflake and normal flakes also put `outPath` in the store on way or another.

roberth reviewed (commented):
Seems useful.

Some thoughts in the file comments +

- please move this to a new `lib/flakes.nix` file. Trivial is the worst name. Might as well call it utils ;)
- add one or two tests in `lib/tests`
- format the new file with `nixpkgs-fmt`

MatthewCroughan committed:
lib: init flakes.nix

This commit creates flakes.nix, which is a library containing functions
which relate to interacting with flakes. It also moves related functions
from trivial.nix into it.

MatthewCroughan commented:
@roberth I've done as you requested. All that's left to do is add a test in lib/tests. I will do this shortly.

MatthewCroughan committed:
lib/tests: add test for callLocklessFlake

MatthewCroughan committed:
lib/tests: use subflake to test callLocklessFlake

MatthewCroughan committed:
lib/tests: evaluate value from subflake with callLocklessFlake

MatthewCroughan commented:
I've added what I think is a decent test, I'm not sure how i could do it better, so please advise on a better way to do it if there is one.

adisbladis commented:
I think this is a very bad idea and I have reverted it in https://github.com/NixOS/nixpkgs/pull/173093.

roberth commented:
I figured users of this function would anticipate breaking changes, as Flakes are clearly advertised to be experimental.
I do not think preemptively removing it is particularly productive, but I respect that other maintainers don't want to be responsible for this code.

https://github.com/NixOS/rfcs/pull/82 has been mentioned as a solution in the other thread. I've offered to shepherd it some time ago. If you're interested, please join.

MatthewCroughan commented:
I do not think it's fair to say that Flakes are experimental, given how many users there are, and the plan to stabilize them. They may be unstable, but I do not agree that they are experimental.

roberth commented:
Flakes are deeply flawed when it comes to user facing design/implementation choices. If anything, they should be considered more experimental than they currently are. But maybe I should have more faith in upstream's courage to make (very) breaking changes.

That said, I'll reiterate:
I do not think preemptively removing it is particularly productive.

tomberek commented:
I don't mind the revert for now. I'm working on a better replacement for this that will be strictly speaking more compliant with flakes/subflakes as well as a bit more powerful (allow dynamically generated input + follows attrsets).

7c6f434c commented:
Usually, Nixpkgs cares about a strict subset of stable features of (non-latest, because all the interwoven release cycles) stable Nix. There are useful things easily doable in Nix but not expected to be done in Nixpkgs.

adisbladis commented:
> I do not think preemptively removing it is particularly productive.

Merging it in the first place set a bad precedent with far reaching implications for the stability expectation of all of `lib`.

zimbatm commented:
Since this got reverted, I made a PR to add it to flake-utils. What do you think?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/url-of-flake-inside-flake/32939/2
